compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1098,66 +1098,135 @@ private BType inferTableMemberType(List<BType> memTypes, BType expType) {
+                }
             }
-            fieldNames.add(fieldName);
+        }
 
-            boolean isOptional = true;
-            for (BField commonField : commonFieldSet) {
-                if (commonField.name.value.equals(fieldName)) {
-                    isOptional = false;
-                    requiredFieldNames.add(commonField.name.value);
+        LinkedHashSet<BField> inferredFields = new LinkedHashSet<>();
+        int memTypesSize = memTypes.size();
+
+        fieldNameToFields.forEach((fieldName, fields) -> {
+            List<BType> uniqueTypes = getUniqueTypesFromSameNameFieldList(fields);
+
+            for (BType memType : memTypes) {
+                BRecordType bMemType = (BRecordType) memType;
+                if (bMemType.sealed) {
+                    continue;
+                }
+
+                if (bMemType.fields.containsKey(fieldName)) {
+                    continue;
+                }
+
+                BType restFieldType = bMemType.restFieldType;
+                if (isUniqueType(uniqueTypes, restFieldType)) {
+                    uniqueTypes.add(restFieldType);
                 }
             }
 
+            BField resultantField = replaceFieldType(fields.get(0), uniqueTypes);
+            boolean isOptional = hasOptionalFields(fields) || fields.size() != memTypesSize;
+
             if (isOptional) {
-                field.symbol.flags = Flags.asMask(EnumSet.of(Flag.OPTIONAL));
-            } else if (requiredFieldNames.contains(fieldName) && keySpecifierFieldNames.contains(fieldName)) {
-                field.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED)) + Flags.asMask(EnumSet.of(Flag.READONLY));
-            } else if (requiredFieldNames.contains(fieldName)) {
-                field.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED));
+                resultantField.symbol.flags = Flags.asMask(EnumSet.of(Flag.OPTIONAL));
+            } else if (keySpecifierFieldNames.contains(fieldName)) {
+                resultantField.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED)) +
+                        Flags.asMask(EnumSet.of(Flag.READONLY));
+            } else {
+                resultantField.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED));
+            }
+
+            inferredFields.add(resultantField);
+        });
+
+        return createTableConstraintRecordType(inferredFields, restFieldTypes, tableConstructorExpr.pos);
+    }
+
+    /**
+     * Gives unique {@code BType} list, extracted from {@code BField} list in which all fields having same field name.
+     *
+     * @param fields the list of same name fields
+     * @return a {@code List<BType>}
+     */
+    private List<BType> getUniqueTypesFromSameNameFieldList(List<BField> fields) {
+        List<BType> uniqueTypes = new ArrayList<>();
+
+        for (BField field : fields) {
+            BType type = field.getType();
+            if (type.tag == TypeTags.UNION) {
```

Comment:
We may have to update this change for the type reference type change.

For example,

```ballerina
type Foo int|string;

public function main() {
    Foo f = 1;

    var t = table [
        {a: f},
        {a: 1}
    ];
    int _ = t;
}
```

results in

```
incompatible types: expected 'int', found 'table<record {| (Foo|int) a; |}>'
```

although `Foo` already includes `int`.